### PR TITLE
Selective attention to directory contents

### PR DESCRIPTION
**Pull request by thompsonmj**
Addresses #5 
Also enables the user to have more control over the contents of the directory to crawl and hash through optionally passing files specifying patterns to either ignore or include, but not both.

- To ignore, specify directory and/or file patterns relative to the target directory root that should be ignored.
- Broad patterns may be negated with `!`. E.g. dotfiles are ignored by default, but to explicitly negate something from being ignored, you could add something like `!.git/logs*` to the ignore file to hash files under this directory while still ignoring other files under `.git/` if desired.
- To exclusively include directories or files matching a set of patterns, use an include file. Only patterns matching entries here will be hashed; all others ignored.
- The output file specified is not hashed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,21 +18,34 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+                        Filepath for the include patterns file
 ```
 
 ### Python Package Usage
 ```python
 from sumbuddy import get_checksums
 
-get_checksums("path/to/image/folder", "path/to/checksums.csv")
+input_dir = "path/to/content/dir"
+output_file = "./content_checksums.csv"
```
Comment (egrace479) [inline README.md:42]:
```suggestion
output_file = "path/to/checksums.csv"
```
Stays consistent with the output description below and also more general.

File: src/sumbuddy/__main__.py [hunk 2]
```diff
@@ -2,33 +2,123 @@
...
+    # Handle exclude mode
+    for pattern in patterns:
+        if fnmatch.fnmatch(relative_filepath, pattern):
+            # Pattern matches for ignore, exclude this file
+            return False  
+
+    return not relative_filepath.split(os.sep)[0].startswith('.')
+
+
+def gather_file_paths(input_directory, output_filepath, ignore_file=None, include_file=None):
```
Comment (egrace479) [inline src/sumbuddy/__main__.py:59]: I feel like there are too many functions in this file and we should move some into other helper files so `__main__` isn't so overloaded. Perhaps just keep `get_checksums` and `main` here.

**Comment by thompsonmj** [comment]
Architecture has been rearranged to split into the following:
- Expose 3 functions to the user, `get_checksums` to function as the CLI would, along with `gather_file_paths` in case a user just wants to collect and view the filtered results for their directory and `checksum` to do a checksum on a single file.
- `__main__.py` to handle CLI args and define `get_checksums`
- `filter.py` to handle including/ignoring directory contents based on patterns in include/ignore file
- `mapper.py` to define `gather_file_paths` based on patterns to include/ignore. Option kept here to swap out future filter modules.
- `hasher.py` to define `checksum`, set up to add new hash algorithms (note the output file now contains a `"checksum"` column instead of `"md5"` accordingly)

Also the `tests/` folder was renamed to `examples/`, since we'll be adding unit tests soon, and `examples/` will serve as illustrations primarily of how the filter patterns work.

**Comment by egrace479** [comment]
I really like the idea of recognizing different checksum selections in the output file; however, I was thinking, maybe instead of output being `filepath,filename,checksum` the `checksum` column name is whichever checksum is selected? So `md5` is default, but if `sha256` or something like that is picked, that becomes the column name. I think it would help for clarity because someone may not recall which checksum they ran and this would preserve that information.

**Comment by thompsonmj** [comment]
Got it! Also added the CLI code to enable hash algorithm selection by the user and removed the old stray `md5_checksum` function from `hasher.py`.

File: README.md [hunk 3]
```diff
@@ -18,22 +18,46 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+                        Directory to traverse for files
+  --output-file OUTPUT_FILE
+                        Filepath for the output CSV file
+  --ignore-file IGNORE_FILE
+                        Filepath for the ignore patterns file
+  --include-file INCLUDE_FILE
+                        Filepath for the include patterns file
+  --algorithm ALGORITHM
+                        Hash algorithm to use (default: md5)
 ```
```
Comment (egrace479) [inline README.md:38]:
```suggestion
Note: this package is still under development and currently only supports `md5` checksums.

```
We'll probably want to provide a list of options as we add them, but this is probably sufficient for v0.1.0.

Comment (thompsonmj) [reply README.md:38]: Actually it already works with everything available through `hashlib`. The list can be seen with `hashlib.algorithms_available`.

File: src/sumbuddy/__main__.py [hunk 4]
```diff
@@ -1,34 +1,48 @@
...
 
 def main():
-    parser = argparse.ArgumentParser(description="Generate CSV with filepath, filename, and MD5 checksums for all files in a given directory") 
-    parser.add_argument("--input-dir", required=True, help="Directory to traverse for files") 
+    parser = argparse.ArgumentParser(description="Generate CSV with filepath, filename, and checksums for all files in a given directory")
+    parser.add_argument("--input-dir", required=True, help="Directory to traverse for files")
     parser.add_argument("--output-file", required=True, help="Filepath for the output CSV file")
+    parser.add_argument("--ignore-file", help="Filepath for the ignore patterns file")
+    parser.add_argument("--include-file", help="Filepath for the include patterns file")
+    parser.add_argument("--algorithm", default="md5", help="Hash algorithm to use (default: md5)")
```
Comment (egrace479) [inline src/sumbuddy/__main__.py:38]:
```suggestion
    parser.add_argument("--algorithm", default="md5", help="Hash algorithm to use (default: md5, see options with 'hashlib.algorithms_available')")
```

File: README.md (hunk 3 above)
Comment (egrace479) [reply README.md:38]: Ah, fabulous! See updated suggestion above (also added to help text) :)

File: README.md [hunk 5]
```diff
@@ -18,22 +18,46 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+  --input-dir INPUT_DIR
+                        Directory to traverse for files
+  --output-file OUTPUT_FILE
+                        Filepath for the output CSV file
+  --ignore-file IGNORE_FILE
+                        Filepath for the ignore patterns file
+  --include-file INCLUDE_FILE
+                        Filepath for the include patterns file
+  --algorithm ALGORITHM
+                        Hash algorithm to use (default: md5)
```
Comment (egrace479) [inline README.md:36]:
```suggestion
                        Hash algorithm to use (default: md5, , see options with 'hashlib.algorithms_available')
```

File: src/sumbuddy/__main__.py [hunk 6]
```diff
@@ -1,34 +1,48 @@
...
 
-def md5_checksum(file_path):
-    hash_md5 = hashlib.md5()
-    with open(file_path, "rb") as f:
-        for chunk in iter(lambda: f.read(4096), b""):
-            hash_md5.update(chunk)
-    return hash_md5.hexdigest()
+def get_checksums(input_directory, output_filepath, ignore_file=None, include_file=None, algorithm='md5'):
+    """
+    Generate a CSV file with the filepath, filename, and checksum of all files in the input directory according to patterns to ignore or include.
```
Comment (egrace479) [inline src/sumbuddy/__main__.py:11]:
```suggestion
    Generate a CSV file with the filepath, filename, and checksum of all files in the input directory according to patterns to ignore or include. Checksum column is labeled by the selected algorithm (e.g., 'md5' or 'sha256').
    
    Parameters:
    ------------
    input_directory - String. Directory to traverse for files.
    output_filepath - String. Filepath for the output CSV file.
    ignore_file - String [optional]. Filepath for the ignore patterns file. Dotfiles are ignored by default and output file won't be hashed.
    include_file - String [optional]. Filepath for the include patterns file.
    algorithm - String. Algorithm to use for checksums. Default: 'md5', see options with 'hashlib.algorithms_available'.
```

File: src/sumbuddy/hasher.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import hashlib
+
+class Hasher:
+    def __init__(self, algorithm='md5'):
+        self.algorithm = algorithm
+
+    def checksum(self, file_path, algorithm=None):
+        """
+        Calculate the checksum of a file using the specified algorithm.
```
Comment (egrace479) [inline src/sumbuddy/hasher.py:9]:
```suggestion
        Calculate the checksum of a file using the specified algorithm.
        
        Parameters:
        ------------
        file_path - String. Path to file to apply checksum function.
        algorithm - String. Hash function to use for checksums. Default: 'md5', see options with 'hashlib.algorithms_available'.
        
        Returns:
        ---------
        hashlib.<algorithm>.hexdigest - String. Hash of file.
```

File: src/sumbuddy/mapper.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import os
+from sumbuddy.filter import Filter
+
+class Mapper:
+    def __init__(self, filter_manager=None):
+        self.filter_manager = filter_manager or Filter()
+
+    def gather_file_paths(self, input_directory, ignore_file=None, include_file=None):
+        """
+        Gather file paths in the input directory, applying pattern rules.
```
Comment (egrace479) [inline src/sumbuddy/mapper.py:10]:
```suggestion
        Gather file paths in the input directory, applying pattern rules.
        
        Parameters:
        ------------
        input_directory - String. Directory to traverse for files.
        ignore_file - String [optional]. Filepath for the ignore patterns file. Dotfiles are ignored by default and output file won't be hashed.
        include_file - String [optional]. Filepath for the include patterns file.
        
        Returns:
        ---------
        file_paths - List. Files in input_directory that are included (i.e., files matching  include_file patterns or not matching ignore_file patterns).
```

**Review by egrace479 — COMMENTED** [review]
Could also use `@params` if preferred. The idea is just to make sure the exposed functions have instructions.

File: README.md [hunk 9]
```diff
@@ -18,22 +18,79 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+  --algorithm ALGORITHM
+                        Hash algorithm to use (default: md5, see options with 'hashlib.algorithms_available')
 ```
 
+#### CLI Examples
+You may use the `examples/` directory to test the CLI.
+
+```bash
+sum-buddy --input-dir examples/example_content --output-file examples/checksums.csv --ignore-file examples/.sbignore_all --algorithm md5
+Calculating checksums: 0it [00:00, ?it/s]
```
Comment (egrace479) [inline README.md:44]:
````suggestion
```
Output
```console
Calculating checksums: 0it [00:00, ?it/s]
````

File: README.md [hunk 10]
```diff
@@ -18,22 +18,79 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+Checksums written to examples/checksums.csv
+cat examples/checksums.csv
```
Comment (egrace479) [inline README.md:46]:
````suggestion
```
```bash
cat examples/checksums.csv
```
Output:
```console
````

File: README.md [hunk 11]
```diff
@@ -18,22 +18,79 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+```bash
+sum-buddy --input-dir examples/example_content --output-file examples/checksums.csv --ignore-file examples/.sbignore_all --algorithm md5
+Calculating checksums: 0it [00:00, ?it/s]
+Checksums written to examples/checksums.csv
+cat examples/checksums.csv 
+filepath,filename,md5
+```
+
+```bash
+$ sum-buddy --input-dir examples/example_content --output-file examples/checksums.csv --ignore-file examples/.sbignore_all_except_dots --algorithm md5
```
Comment (egrace479) [inline README.md:51]:
````suggestion
sum-buddy --input-dir examples/example_content --output-file examples/checksums.csv --ignore-file examples/.sbignore_all_except_dots --algorithm md5
```
Output:
```console
````

File: README.md [hunk 12]
```diff
@@ -18,22 +18,79 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+Calculating checksums: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 2178.48it/s]
+Checksums written to examples/checksums.csv
+cat examples/checksums.csv
```
Comment (egrace479) [inline README.md:54]:
````suggestion
```
```bash
cat examples/checksums.csv
```
Output:
```console
````

File: README.md [hunk 13]
```diff
@@ -18,22 +18,79 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+Calculating checksums: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 2178.48it/s]
+Checksums written to examples/checksums.csv
+cat examples/checksums.csv 
+filepath,filename,md5
+examples/example_content/.hidden_file,.hidden_file,d41d8cd98f00b204e9800998ecf8427e
+examples/example_content/.hidden_dir/.hidden_file,.hidden_file,d41d8cd98f00b204e9800998ecf8427e
+examples/example_content/.hidden_dir/file.txt,file.txt,7d52c7437e9af58dac029dd11b1024df
+```
+
+You may explore the filtering capabilities of the `--ignore-file` and `--include-file` options by using the provided example files under `examples/` and the CSV output files in `examples/expected_outputs/`.
```
Comment (egrace479) [inline README.md:61]:
```suggestion
You may explore the filtering capabilities of the `--ignore-file` and `--include-file` options by using the provided example files under `examples/` and pointing at `examples/example_content`. The expected CSV output files are provided in `examples/expected_outputs/`.
```

**Review by egrace479 — CHANGES_REQUESTED** [review]
Also, add the bullet points describing the PR changes at line 17 (`How it Works`).

**Comment by thompsonmj** [comment]
To make the tool more standardized, filtering behavior has been adjusted to behave just like `git` treats `.gitignore`. This way, users will not need to learn new behavior.

A convenience argument `--ignore-hidden` has also been added to indicate when hidden files/directories should be ignored without needing to create a file specifying patterns.

File: src/sumbuddy/mapper.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        root_directory = os.path.abspath(input_directory)
+
+        for root, dirs, files in os.walk(input_directory):
+            for name in files:
+                file_path = os.path.join(root, name)
+                if self.filter_manager.should_include(file_path, root_directory):
+                    file_paths.append(file_path)
+
+        return file_paths
+
```
Comment (egrace479) [inline src/sumbuddy/mapper.py:51]:
```suggestion

```

File: src/sumbuddy/__main__.py [hunk 15]
```diff
@@ -1,34 +1,55 @@
...
+    print(f"Checksums written to {output_filepath}")
 
 def main():
-    parser = argparse.ArgumentParser(description="Generate CSV with filepath, filename, and MD5 checksums for all files in a given directory") 
-    parser.add_argument("--input-dir", required=True, help="Directory to traverse for files") 
+    parser = argparse.ArgumentParser(description="Generate CSV with filepath, filename, and checksums for all files in a given directory")
+    parser.add_argument("--input-dir", required=True, help="Directory to traverse for files")
     parser.add_argument("--output-file", required=True, help="Filepath for the output CSV file")
+    parser.add_argument("--ignore-file", help="Filepath for the ignore patterns file")
+    parser.add_argument("--ignore-hidden", action="store_true", help="Ignore hidden files")
```
Comment (egrace479) [inline src/sumbuddy/__main__.py:44]:
```suggestion
    parser.add_argument("--ignore-hidden", action="store_true", help="Ignore hidden files (Use either this or --ignore-file. Identical to providing an ignore file with '.*')")
```

File: src/sumbuddy/mapper.py (hunk 14 above)
Comment (egrace479) [inline src/sumbuddy/mapper.py:51]:
```suggestion

```

File: src/sumbuddy/mapper.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        ------------
+        ignore_file - String [optional]. Filepath for the ignore patterns file.
+        ignore_hidden - Boolean [optional]. Whether to ignore hidden files.
+        """
+        self.filter_manager = Filter()
+        if ignore_file:
+            self.filter_manager.read_patterns(filepath=ignore_file)
+        elif ignore_hidden is True:
+            self.filter_manager.read_patterns(ignore_hidden=True)
+        elif ignore_hidden is False:
```
Comment (egrace479) [inline src/sumbuddy/mapper.py:22]:
```suggestion
        else:
```

File: examples/run_examples.bash
Comment (egrace479) [inline examples/run_examples.bash:1]: Not used (as discussed offline), so remove.

**Review by egrace479 — COMMENTED** [review]
See suggestions/comments. There seem to be some extraneous files.

File: examples/expected_outputs/ignore_nothing.csv
[the remark this answers is not in the document]
Comment (thompsonmj) [reply examples/expected_outputs/ignore_nothing.csv:1]: Probably don't _need_, but I think there's value in showing that functionally equivalent commands yield identical outputs.

File: src/sumbuddy/mapper.py (hunk 14 above)
Comment (egrace479) [reply src/sumbuddy/mapper.py:51]: It has a few spaces/tab at the end, so it doesn't register as an end of file line. The suggestion was reading it oddly, and I'm not sure why it duplicated.

**Review by egrace479 — COMMENTED** [review]
See inline suggestions and one question.

Great update!

File: src/sumbuddy/__main__.py [hunk 17]
```diff
@@ -1,34 +1,75 @@
...
+
+    available_algorithms = ', '.join(hashlib.algorithms_available)
+        
+    parser = argparse.ArgumentParser(description="Generate CSV with filepath, filename, and checksums for all files in a given directory")
+    parser.add_argument("input_dir", help="Directory to traverse for files")
+    parser.add_argument("-o", "--output-file", help="Filepath for the output CSV file; defaults to stdout", default=None)
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument("-i", "--ignore-file", help="Filepath for the ignore patterns file")
+    group.add_argument("-H", "--include-hidden", action="store_true", help="Include hidden files")
+    parser.add_argument("-a", "--algorithm", default="md5", help=f"Hash algorithm to use (default: md5; available: {available_algorithms})")
```
[the remark this answers is not in the document]
Comment (thompsonmj) [reply src/sumbuddy/__main__.py:60]: I tried `choices`, and it is messy like you mention. I think the error message is informative enough without adding additional checks.

File: examples/expected_outputs/ignore_nothing.csv
[the remark this answers is not in the document]
Comment (thompsonmj) [reply examples/expected_outputs/ignore_nothing.csv:1]: Now, "ignore nothing" works the same as passing `--include-hidden`, so this shows the equivalence of those two behaviors.

File: README.md [hunk 18]
```diff
@@ -18,22 +18,136 @@ pip install git+https://github.com/Imageomics/sum-buddy
...
+  -H, --include-hidden  Include hidden files
+  -a ALGORITHM, --algorithm ALGORITHM
+                        Hash algorithm to use (default: md5; available: ripemd160, sha3_224, sha512_224, blake2b, sha384, sha256, sm3, sha3_256, shake_256, sha512, sha1, sha224, md5, md5-sha1, sha3_384, sha3_512, sha512_256, shake_128, blake2s)
+```
+
+#### CLI Examples
+
+- **Basic Usage:**
+```bash
+$ sum-buddy examples/example_content/
```
[the remark this answers is not in the document]
Comment (thompsonmj) [reply README.md:43]: Weird, thought I put that into a batch ...

**Review by egrace479 — APPROVED** [review]

**Review by egrace479 — APPROVED** [review]

**Pull request merged**